### PR TITLE
Brian Greeson user story 18

### LINKED ISSUE #21: User Story 18, Items Index Page Statistics
**This is the main "catalog" page of the entire site where users will start their e-commerce experience. Visitors to the site, and regular users, will be able to view an index page of all items available for purchase and some basic statistics. Each item will also have a "show" page where more information is shown.**

As any kind of user on the system
When I visit the items index page ("/items")
I see an area with statistics:
- the top 5 most popular items by quantity purchased, plus the quantity bought
- the bottom 5 least popular items, plus the quantity bought

"Popularity" is determined by total quantity of that item ordered

### PR DESCRIPTION
**Pull request by brian-greeson**
# Items index page shows most and least popular items
Closes: #21 
## Added model helper methods
- most_popular(int) 
   - list the (int) number of most popular items ordered by # of items sold(most->least)
- least_popular(int) 
   - list the (int) number of least popular items ordered by # of items sold (least->most)
- tested model methods
## Items index view
- test items index view for most and least popular sections
- test for total quantity of item sold next to each item
## Future considerations
- move item list iteration into partial to dry up code

### CODE REVIEW

File: app/models/item.rb [hunk 1]
```diff
@@ -25,4 +25,20 @@ def no_orders?
     item_orders.empty?
   end
 
+  def self.most_popular(num_items_to_show)
```
Comment (fredrondina) [inline app/models/item.rb:28]: great job with these methods, they look very clean

File: app/views/items/index.html.erb [hunk 2]
```diff
@@ -21,5 +21,49 @@
         <p>Inactive</p>
       <% end %>
     </section>
-    <% end %>
+  <% end %>
+</section>
```
Comment (fredrondina) [inline app/views/items/index.html.erb:25]: this looks great, awesome job doing this solo

File: spec/features/items/index_spec.rb [hunk 3]
```diff
@@ -57,6 +56,57 @@
...
+      item_order_10 = order_1.item_orders.create!(item: tire1, price: tire1.price, quantity: 10)
+      item_order_5 = order_2.item_orders.create!(item: pull_toy5, price: pull_toy5.price, quantity: 5)
+      item_order_4 = order_2.item_orders.create!(item: pull_toy4, price: pull_toy4.price, quantity: 4)
+      item_order_1 = order_1.item_orders.create!(item: pull_toy1, price: pull_toy1.price, quantity: 2)
+
+      visit '/items'
+
+      within "article.popular-items" do
+        expect(tire2.name).to appear_before(tire5.name, only_text: true)
+        expect(tire5.name).to appear_before(tire4.name, only_text: true)
```
Comment (fredrondina) [inline spec/features/items/index_spec.rb:89]: Good call using orderly for these tests

**Review by fredrondina — APPROVED** [review]
Fantastic work

File: app/models/item.rb [hunk 4]
```diff
@@ -25,4 +25,20 @@ def no_orders?
...
+        .order("sum(item_orders.quantity) DESC")
+        .limit(num_items_to_show)
+  end
+
+  def self.least_popular(num_items_to_show)      
+    Item.joins(:item_orders)
+        .select('items.*, sum(item_orders.quantity)')
+        .group(:id)
+        .order("sum(item_orders.quantity) ASC")
+        .limit(num_items_to_show)
```
Comment (kelshadarby) [inline app/models/item.rb:41]: Very easy to read. Nice job with these queries!

File: app/models/item.rb [hunk 5]
```diff
@@ -25,4 +25,20 @@ def no_orders?
...
+  end
+
```
Comment (kelshadarby) [inline app/models/item.rb:43]: It looks like these two queries are the same except for the ASC and DESC, maybe they could be combined and take an "asc_or_desc" argument or something similar to DRY up the code?

File: app/views/items/index.html.erb [hunk 6]
```diff
@@ -21,5 +21,49 @@
...
+<section class="grid-container">
+  <article class="popular-items">
+  <h1>Most Popular</h1>
+  <% @items.most_popular(5).each do |item| %>
+    <section class = "grid-item" id= 'item-<%=item.id%>'>
+      <h2> <%=link_to item.name, "/items/#{item.id}" %> </h2>
+      <p>Sold by: <%=link_to item.merchant.name, "/merchants/#{item.merchant.id}" %></p>
+      <img src= <%= item.image %>>
+      <p> <%= item.description unless @merchant%> </p>
```
Comment (kelshadarby) [inline app/views/items/index.html.erb:34]: US 17 requires us to have the items images on the index page be links to the show page, should we add that here?

File: spec/features/items/index_spec.rb [hunk 7]
```diff
@@ -57,6 +56,57 @@
...
+      pull_toy3 = @brian.items.create(name: "Pull Toy3", description: "Great pull toy!", price: 10, image: "https://tinyurl.com/rzhm3qd", inventory: 32)
+      pull_toy4 = @brian.items.create(name: "Pull Toy4", description: "Great pull toy!", price: 10, image: "https://tinyurl.com/rzhm3qd", inventory: 32)
+      pull_toy5 = @brian.items.create(name: "Pull Toy5", description: "Great pull toy!", price: 10, image: "https://tinyurl.com/rzhm3qd", inventory: 32)
+      tire1 = @meg.items.create(name: "Gatorskins1", description: "They'll never pop!", price: 100, image: "https://tinyurl.com/tn7jnts", inventory: 12)
+      tire2 = @meg.items.create(name: "Gatorskins2", description: "They'll never pop!", price: 100, image: "https://tinyurl.com/tn7jnts", inventory: 12)
+      tire3 = @meg.items.create(name: "Gatorskins3", description: "They'll never pop!", price: 100, image: "https://tinyurl.com/tn7jnts", inventory: 12)
+      tire4 = @meg.items.create(name: "Gatorskins4", description: "They'll never pop!", price: 100, image: "https://tinyurl.com/tn7jnts", inventory: 12)
```
Comment (kelshadarby) [inline spec/features/items/index_spec.rb:69]: Nice job on this test! Great work!!

File: spec/features/users/show_spec.rb [hunk 8]
```diff
@@ -1,41 +1,39 @@
...
+  before(:all) do
+      @user1 = User.new(
+        email_address: "user1@example.com",
+        password: "password",
+        user_detail: UserDetail.new(
+          name: "User 1",
+          street_address: "123 Example St",
+          city: "Userville",
+          state: "State 1",
+          zip_code: "12345"
```
Comment (kelshadarby) [inline spec/features/users/show_spec.rb:13]: Nice use of before all!

File: spec/models/item_spec.rb [hunk 9]
```diff
@@ -48,4 +48,48 @@
...
+      tire6 = @bike_shop.items.create(name: "Gatorskins6", description: "They'll never pop!", price: 100, image: "https://tinyurl.com/tn7jnts", inventory: 6)
+      order_1 = Order.create!(name: 'Meg1', address: '123 Stang Ave', city: 'Hershey', state: 'PA', zip: 17033)
+      order_2 = Order.create!(name: 'Meg2', address: '123 Stang Ave', city: 'Hershey', state: 'PA', zip: 17033)
+      item_order_2 = order_1.item_orders.create!(item: tire2, price: tire2.price, quantity: 20)
+      item_order_7 = order_2.item_orders.create!(item: tire1, price: tire1.price, quantity: 10)
+      item_order_1 = order_1.item_orders.create!(item: tire1, price: tire1.price, quantity: 1)
+      item_order_6 = order_2.item_orders.create!(item: tire6, price: tire1.price, quantity: 6)
+      item_order_5 = order_2.item_orders.create!(item: tire5, price: tire5.price, quantity: 5)
+      item_order_4 = order_2.item_orders.create!(item: tire4, price: tire4.price, quantity: 4)
+      item_order_3 = order_1.item_orders.create!(item: tire3, price: tire3.price, quantity: 3)
```
Comment (kelshadarby) [inline spec/models/item_spec.rb:69]: Nice testing here!!

**Review by kelshadarby — APPROVED** [review]

**Pull request merged**